charger/peblar.go

```diff
@@ -0,0 +1,282 @@
+	peblarRegCpState = 30110
+)
+
+func init() {
+	registry.Add("peblar", NewPeblarFromConfig)
+}
+
+//go:generate go run ../cmd/tools/decorate.go -f decoratePeblar -b *Peblar -r api.Charger -t "api.PhaseSwitcher,Phases1p3p,func(int) error" -t "api.PhaseGetter,GetPhases,func() (int, error)"
+
+// NewPeblarFromConfig creates a Peblar charger from generic config
+func NewPeblarFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := modbus.TcpSettings{
+		ID: 255,
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewPeblar(cc.URI, cc.ID)
+}
+
+// NewPeblar creates Peblar charger
+func NewPeblar(uri string, id uint8) (api.Charger, error) {
+	conn, err := modbus.NewConnection(uri, "", "", 0, modbus.Tcp, id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Register contains the physically connected phases
+	b, err := conn.ReadInputRegisters(peblarRegPhaseCount, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	wb := &Peblar{
+		conn:   conn,
+		curr:   6000,                       // assume min current
+		phases: binary.BigEndian.Uint16(b), // required for retrieving the right amount of voltage/current registers
+	}
+
+	b, err = conn.ReadInputRegisters(peblarRegIndepRelay, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	var phasesS func(int) error
+	var phasesG func() (int, error)
+
+	if binary.BigEndian.Uint16(b) == 1 {
+		phasesS = wb.phases1p3p
+		phasesG = wb.getPhases
+	}
+
+	return decoratePeblar(wb, phasesS, phasesG), err
+}
+
+// Status implements the api.Charger interface
+func (wb *Peblar) Status() (api.ChargeStatus, error) {
+	b, err := wb.conn.ReadInputRegisters(peblarRegCpState, 1)
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	switch s := rune(encoding.Uint16(b)); s {
+	case 'A', 'B', 'C':
+		return api.ChargeStatus(s), nil
+	default:
+		return api.StatusNone, fmt.Errorf("invalid status: %d", s)
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (wb *Peblar) Enabled() (bool, error) {
+	return verifyEnabled(wb, wb.enabled)
+}
+
+// Enable implements the api.Charger interface
+func (wb *Peblar) Enable(enable bool) error {
+	var current uint32
+	if enable {
+		current = wb.curr
+	}
+
+	err := wb.setCurrent(current)
+	if err == nil {
+		wb.enabled = enable
+	}
+
+	return err
+}
+
+// setCurrent writes the current limit in mA
+func (wb *Peblar) setCurrent(current uint32) error {
+	b := make([]byte, 4)
+	binary.BigEndian.PutUint32(b, current)
+
+	_, err := wb.conn.WriteMultipleRegisters(peblarRegModbusCurrentLimit, 2, b)
+	return err
+}
+
+// MaxCurrent implements the api.Charger interface
+func (wb *Peblar) MaxCurrent(current int64) error {
+	return wb.MaxCurrentMillis(float64(current))
+}
+
+var _ api.ChargerEx = (*Peblar)(nil)
+
+// MaxCurrent implements the api.ChargerEx interface
+func (wb *Peblar) MaxCurrentMillis(current float64) error {
+	if current < 6 {
+		return fmt.Errorf("invalid current %.1f", current)
+	}
+
+	err := wb.setCurrent(wb.curr)
```

Comment:
doh :O

Reply:
Well that resulted in unexpected behavior..